File: 2024/q15.py
#!/usr/bin/env python
import io
import string
from collections import deque


def display(grid, path=[], /, pois=None, color=None):
    palette = ['37;40', '37;41', '30;42', '30;43', '37;44', '37;45']
    c = palette[(color or 0) % len(palette)]
    coff = '\033[0m'
    con = f'\033[{c}m'
    xon = f'\033[37;41m'
    minx = min(x for y,x in grid)
    maxx = max(x for y,x in grid)
    miny = min(y for y,x in grid)
    maxy = max(y for y,x in grid)
    with io.StringIO() as so:
        for y in range(miny, maxy+1):
            for x in range(minx, maxx+1):
                c = grid.get((y,x), ' ')
                if pois and (y,x) in pois:
                    s = f'{xon}{c}{coff}'
                elif (y,x) in path:
                    s = f'{con}{c}{coff}'
                else:
                    s = c
                print(s, end='', file=so)
            print(file=so)
        print(so.getvalue())


def part1(data):
    grid = {(y,x):c for y,s in enumerate(data.strip().splitlines())
        for x,c in enumerate(s) if c in '.H'}
    herbs = [p for p,c in grid.items() if c == 'H']
    start = next((y,x) for (y,x),c in grid.items() if y == 0)
    neib = [(1,0), (-1,0), (0,1), (0,-1)]
    fringe = deque([(0, start)])
    seen = set()
    while fringe:
        dist, pos = fringe.popleft()
        if pos in herbs:
            return 2 * dist
        if pos in seen: continue
        seen.add(pos)
        y,x = pos
        for dy,dx in neib:
            q = (y+dy, x+dx)
            if q in grid:
                fringe.append((dist+1, q))


def part2(data):
    world = {(y,x):c for y,s in enumerate(data.strip().splitlines())
        for x,c in enumerate(s)}
    grid = {p:c for p,c in world.items() if c not in '#~'}
    herbs = [p for p,c in grid.items() if c != '.']
    start = next((y,x) for (y,x),c in grid.items() if y == 0)
    goal = ''.join(sorted({grid[p] for p in herbs}))
    neib = [(1,0), (-1,0), (0,1), (0,-1)]

    fringe = deque([(0, start, '')])
    seen = set()
    while fringe:
        dist, pos, bag = fringe.popleft()
        if pos == start and bag == goal:
            ans = dist
            break
        k = (pos, bag)
        if k in seen: continue
        seen.add(k)
        y,x = pos
        for dy,dx in neib:
            q = (y+dy, x+dx)
            if q not in grid: continue
            if q in herbs:
                c = grid[q]
                if c not in bag:
                    b = ''.join(sorted(bag + c))
                    fringe.append((dist+1, q, b))
                    continue
            fringe.append((dist+1, q, bag))
    return ans


def unmask(v, m):
    n = (v + m).bit_length()
    f = ((1 << n) - 1) - m
    return v & f


def part3(data):
    world = {(y,x):c for y,s in enumerate(data.strip().splitlines())
        for x,c in enumerate(s)}
    grid = {p:c for p,c in world.items() if c not in '#~'}
    start = next((y,x) for (y,x),c in grid.items() if y == 0)
    neib = [(1,0), (-1,0), (0,1), (0,-1)]

    def walk(grid, seen, start, color=1):
        wave = list()
        fringe = deque([start])
        while fringe:
            pos = fringe.popleft()
            if pos in seen: continue
            seen.add(pos)
            y,x = pos
            for dy,dx in neib:
                q = (y+dy, x+dx)
                if (c := grid.get(q)) is not None:
                    if c:
                        wave.append(q)
                    else:
                        fringe.append(q)
        res = 0
        ls = 0
        for w in wave:
            t,l = walk(grid, seen, w, color=color+1)
            res += t
            ls |= unmask(l, grid[start])

        fringe = deque([(0, start, ls)])
        seen = {(start, ls)}
        while fringe:
            dist, pos, bag = fringe.popleft()
            if pos == start and not bag:
                res += dist
                break
            y,x = pos
            for dy,dx in neib:
                q = (y+dy, x+dx)
                if (c := grid.get(q)) is not None:
                    b = unmask(bag, c)
                    k = (q, b)
                    if k in seen: continue
                    seen.add(k)
                    fringe.append((dist+1, q, b))

        grid[start] |= ls
        return res, grid[start]

    abc = string.ascii_uppercase
    masked = {p:(0 if c == '.' else (1 << abc.index(c))) for p,c in grid.items()}
    ans,_ = walk(masked, set(), start)
    return ans


data = """
#####.#####
#.........#
#.######.##
#.........#
###.#.#####
#H.......H#
###########
"""
assert part1(data) == 26

data = """
##########.##########
#...................#
#.###.##.###.##.#.#.#
#..A#.#..~~~....#A#.#
#.#...#.~~~~~...#.#.#
#.#.#.#.~~~~~.#.#.#.#
#...#.#.B~~~B.#.#...#
#...#....BBB..#....##
#C............#....C#
#####################
"""
assert part2(data) == 38
assert part3(data) == 38


data = open('q15_p1.txt').read()
ans = part1(data)
print('part1:', ans)

data = open('q15_p2.txt').read()
ans = part2(data)
print('part2:', ans)

data = open('q15_p3.txt').read()
ans = part3(data)
print('part3:', ans)

File: 2024/q15.jl
#!/usr/bin/env julia

const Pos = CartesianIndex{2}


function parsenotes(text)
    grid = stack(split(text), dims=1)
    grid[grid .== '~'] .= '#'
    i = findfirst(==('.'), grid[1,:])
    return (grid, Pos(1, i))
end


function part1_(data)
    grid, start = parsenotes(data)
    goal = findall(==('H'), grid)
    neibs = [(0,1),(0,-1),(1,0),(-1,0)] .|> Pos
    fringe = [(0, start)]
    seen = Set{Pos}()
    for (w, p) in fringe
        p in seen && continue
        push!(seen, p)
        p in goal && return 2w
        for q in p .+ neibs
            if get(grid, q, '#') != '#'
                push!(fringe, (w+1, q))
            end
        end
    end
end


function solve_(data; cutoff=false)
    grid, start = parsenotes(data)
    names = data ∩ ('A':'Z')
    n = length(names)
    goal = 0
    m = zeros(Int, size(grid))
    m[grid .== '#'] .= -1
    for i in findall(in(names), grid)
        goal |= (m[i] = 1<<(grid[i]-'A'))
    end
    neibs = [(0,1),(0,-1),(1,0),(-1,0)] .|> Pos
    fringe = [(0, start, 0)]
    seen = Set{Tuple{Pos,Int}}()
    best = Dict{Int,Int}()
    while !isempty(fringe)
        w, p, t = popfirst!(fringe)
        if t == goal
            p == start && return w
        end
        if cutoff
            u = count_ones(t)
            u + 2 < get!(best, w, u) && continue
            best[w] = max(best[w], u)
        end
        k = (p, t)
        k in seen && continue
        push!(seen, k)
        for q in p .+ neibs
            x = get(m, q, -1)
            if (x > 0) && (x & t == 0)
                push!(fringe, (w+1, q, t|x))
            elseif x != -1
                push!(fringe, (w+1, q, t))
            end
        end
    end
end


function solve(data)
    grid, start = parsenotes(data)
    names = data ∩ ('A':'Z')
    neibs = [(0,1),(0,-1),(1,0),(-1,0)] .|> Pos

    function walk(grid, start, seen)
        wave = Pos[]
        fringe = [start]
        while !isempty(fringe)
            pos = popfirst!(fringe)
            pos in seen && continue
            push!(seen, pos)
            for q in pos .+ neibs
                x = get(grid, q, -1)
                if x > 0
                    push!(wave, q)
                elseif x == 0
                    push!(fringe, q)
                end
            end
        end
        path, avail = 0, 0
        for p in wave
            n,i = walk(grid, p, seen)
            path += n
            avail |= i
        end
        fringe2 = [(0, start, avail & ~grid[start])]
        seen2 = Set{Tuple{Pos,Int}}()
        while !isempty(fringe2)
            dist, pos, bag = popfirst!(fringe2)
            if (bag == 0) && (pos == start)
                path += dist; break
            end
            for q in pos .+ neibs
                x = get(grid, q, -1)
                x == -1 && continue
                t = bag & ~x
                k = (q, t)
                k in seen2 && continue
                push!(seen2, k)
                push!(fringe2, (dist+1, q, t))
            end
        end
        grid[start] |= avail
        return (path, grid[start])
    end

    m = zeros(Int32, size(grid))
    m[grid .== '#'] .= -1
    for i in findall(∈(names), grid)
        m[i] = 1 << (grid[i]-'A')
    end
    walk(m, start, Set{Pos}()) |> first
end


part1 = solve
part2 = solve
part3 = solve


data = raw"
#####.#####
#.........#
#.######.##
#.........#
###.#.#####
#H.......H#
###########
"
@assert part1(data) == 26


data = raw"
##########.##########
#...................#
#.###.##.###.##.#.#.#
#..A#.#..~~~....#A#.#
#.#...#.~~~~~...#.#.#
#.#.#.#.~~~~~.#.#.#.#
#...#.#.B~~~B.#.#...#
#...#....BBB..#....##
#C............#....C#
#####################
"
@assert part2(data) == 38


data = readchomp("q15_p1.txt")
println("part1: ", part1(data))
data = readchomp("q15_p2.txt")
println("part2: ", part2(data))
data = readchomp("q15_p3.txt")
println("part3: ", part3(data))
